packages/devextreme/testing/tests/DevExpress.ui.widgets/tabs.tests.js

```diff
@@ -422,6 +422,100 @@ QUnit.module('General', () => {
+                items: [
+                    { text: '0' },
+                    { text: '1' },
+                    { text: '2' },
+                    { text: '3' }
+                ],
+                onSelectionChanging: this.onSelectionChangingSpy,
+                onSelectionChanged: this.onSelectionChangedSpy,
+                selectedIndex: 0
+            });
+            this.tabInstance = this.$tabs.dxTabs('instance');
+        },
+        afterEach() {
+            this.clock.restore();
+        }
+    }, () => {
+        QUnit.test('cancelling selection event synchronously', function(assert) {
+            this.onSelectionChangingSpy = sinon.spy(function(e) {
+                e.cancel = true;
+            });
+
+            this.tabInstance.option('onSelectionChanging', this.onSelectionChangingSpy);
+            this.$item = this.$tabs.find('.dx-tab').eq(1);
+            this.$item.trigger('dxclick');
+
+            assert.strictEqual(this.onSelectionChangingSpy.callCount, 1, 'onSelectionChanging should be called');
+            assert.strictEqual(this.onSelectionChangedSpy.callCount, 0, 'onSelectionChanged should not be called');
+            assert.ok(this.onSelectionChangingSpy.getCall(0).args[0].cancel, 'e.cancel should be set to true');
+            assert.strictEqual(this.tabInstance.option('selectedIndex'), 0, 'selectedIndex should remain 0 when cancelled');
+        });
+
+        QUnit.test('cancelling selection event asynchronously', function(assert) {
+            this.onSelectionChangingSpy = sinon.spy(function(e) {
+                e.cancel = new Promise((resolve) => {
+                    setTimeout(() => {
+                        resolve(true);
+                    }, 5000);
+                });
+            });
+
+            this.tabInstance.option('onSelectionChanging', this.onSelectionChangingSpy);
+            this.$item = this.$tabs.find('.dx-tab').eq(1);
+            this.$item.trigger('dxclick');
+
+            this.clock.tick(6000);
+            this.onSelectionChangingSpy.getCall(0).args[0].cancel.then((cancel) => {
+                assert.strictEqual(this.onSelectionChangingSpy.callCount, 1, 'onSelectionChanging should be called');
+                assert.strictEqual(this.onSelectionChangedSpy.callCount, 0, 'onSelectionChanged should not be called');
+                assert.ok(cancel, 'e.cancel should be set to true');
+                assert.strictEqual(this.tabInstance.option('selectedIndex'), 0, 'selectedIndex should remain 0 when cancelled');
+            });
+            this.clock.tick(6000);
+        });
+
+        QUnit.test('default to false when e.cancel is not updated', function(assert) {
+
+            this.tabInstance.option('onSelectionChanging', this.onSelectionChangingSpy);
+            this.$item = this.$tabs.find('.dx-tab').eq(1);
+            this.$item.trigger('dxclick');
+
+            assert.strictEqual(this.onSelectionChangingSpy.callCount, 2, 'onSelectionChanging should be called');
```

Comment:
I think this is correct anton... the reason its 2 because when in initialization our collection widget automatically runs selection procedure in tabs u can see it on initMarkup method when initializing our markup its also initialize selection in which it runs onSelectionChanged method that runs our onSelectionChanging first always the reason its 2 because the second call happens during clicking of next item which calls again the same steps as initialization in which it triggers our second call for onSelectionChanging  u can start to look or debug in selectItem method in collection edit. 


P.S:
based from my research when running _normalizeSelectedItems method it calls setSelection method which triggers on our selection API selectedItemKeys which it runs again onSelectionChanged method...

Reply:
[bebf086](https://github.com/DevExpress/DevExtreme/pull/27860/commits/bebf0867b9c1214b545d0d2fb3c7c46c2e3ff9c8)